Protobuf

eeberhard opened the pull request:
Gentlemen, the time has come. I present a feature that I have thought about for a long time and finally have ready to review.

## A quick summary of the design:

Things are heavily based around Google protobuf. Maybe check the overview before digging into this PR just so you know roughly what to expect: https://developers.google.com/protocol-buffers

### proto files
The .proto files define the messages. It's very much like the structures defined in franka_lwi_proto, except using google syntax. There is inheritance wherever possible, to allow nested messages to reuse elements. For example, State and SpatialState are heavily reused.

At the "top level" of the .proto messages is a StateMessage definition. This has a field with the special `oneof` property. It essentially acts as a C union / C++ std::variant - it can contain one of the listed messages. This allows all messages to be sent as the StateMessage type and enables a reflection API to see what the message content actually is.

### proto bindings

The .proto files themselves just define the message structure, but are not readable by code right away. For this, the protobuf compiler (protoc) is invoked on the files to automatically generate cpp and python bindings (as well as go, dart, java, ...). This turns the structures into classes with various access methods for the specified types. In addition, and most importantly for our purposes, the generated classes have the methods`SerializeToString` and `ParseFromString` , which respectively encode and decode between the object data and a compact, efficient "wire format", a serialized binary representation of the data. 

With the bindings in hand, we can easily create instances of these message types, set their fields, convert them to wire format and transmit them to another recipient who uses the same proto definitions and generated bindings to decode them back into messages. Easy!

### clproto

Well, the job is not done yet. Translating between the message type and the wire format is done for us, but we also need to be able to translate between the message type and their respective control libraries types! For example, a `proto::CartesianState` may have all the same data fields as a `state_representation::CartesianState`, but it can't just be cast or converted in one go. For this, a user-facing library is created (control libraries protocol) which implements a translation for each message type to the respective control library type. In most cases, translation code can be re-used, but it's still an unavoidable slew of implementation code (just like the python bindings).

By using template declarations in the `clproto.h` header file, and doing all of the actual translation implementation in template specialisations in the `clproto.cpp` source file, the API can map directly between wire format to control library types, hiding the actual proto types and auto-generated binding code. 

## Trying it yourself

### a note on protobuf dependencies 

The protobuf compiler `protoc` is needed to build the bindings, but is not used after that step. I haven't yet figured out how to build just the runtime library `libprotobuf`, which is the only part that would be needed to build and use `clproto` provided the generated bindings already exist. For other languages, the runtime seems to be more cleanly separated. I mention this only because it takes quite a while to compile `protoc` from source (something like 20 minutes).

### cmake and tests

The cmakelists adds a target to generate the bindings from the proto, but does not automatically invoke it as no other target depends explicitly on that target (even if the library depends on the _files_ the target creates). For this you have to run `make generate_proto_bindings` in the clproto build directory after configuring cmake to generate cpp bindings, or more easily, run `make cpp_bindings` or `make all` in the protobuf directory.

It also adds an intermediate library `clproto_bindings` which is just a convenient way to bundle the auto-generated proto bindings. It is not intended for the end-user.

Using the intermediate library, `clproto` is built and can be installed (though I haven't tested / refined the installation process yet).

Finally a `test_clproto` target is added to test the important stuff with gtest.

### check it out for yourself

If you run `bash run.sh` from within the protocol directory, you'll build and start the interactive container (again, will take a long time if you need to build protobuf). From there, try navigating to `/home/remote/protocol/clproto_cpp`, doing `mkdir build && cd build` and then `cmake  -DBUILD_TESTING=ON ..`. Finally, you should be able to simply `make` to build all the registered targets (the intermediate binding library, the actual clproto library, and the test executable. Run the tests by executing `./test_clproto` from within the build directory 

## Closing thoughts

This PR is slightly over the recommended size of 250 new lines. But, it's one of those things that is the easiest to demonstrate when all the pieces are working. I wrote tests to check the encoding / decoding of our main types, and they all pass. I strongly recommend you hide the `.pb.cc`, `.pb.h` and `_pb2.py` files using the file filter in the review tab. @domire8 I'm sorry to leap ahead so unfairly in our contribution race by adding the generated binding files. My justification is that protoc takes ages to install and shouldn't be necessary for the end-user, so I've added them "pre-compiled for your convenience".

You might also notice I didn't use `data()` / `set_data()` methods and instead used specific accessors and setters. While the proto message structure and the resulting translation would have been much simpler to just contain a vector<double> / repeating double field, I don't think that serves the longer term picture. If we can easily pull out position x, y, z and other low-level structure from the message, it means something like a javascript function in the front-end or a go routine in a database backend can deal directly with the proto types in a sensible and human readable way, even when control libraries itself is not available. For example, a 7-element pose will always remain ambiguous as to the ordering of the respective coefficients, and the representation of the orientation, if you don't have access to the encoding scheme used. Instead, explicit position and quaternion fields are present.

Anyway, all that said, this was a fun 2-day project! Especially the part on Thursday afternoon when I accidentally mounted a volume at the same time as having a Deployment configuration syncing directories over SSH - resulting in the entire project directory being instantly erased from both the container and my host. I had staged a few things on GIT but far from everything, so I lost about 5 hours of work and a lot of motivation. But, after a nice 6pm swim in the lake and a comforting dinner, I got back to it with renewed vigour!

domire8 commented on protocol/protobuf/proto/state_representation/space/cartesian/cartesian_state.proto:
```diff
@@ -0,0 +1,46 @@
+
+message CartesianTwist {
+  SpatialState spatial_state = 1;
+  Vector3d linear_velocity = 4;
```
Are those fields number 4 and 5 on purpose (if so why) or is it from copying? same in CartesianWrench

eeberhard replied:
great catch! thank you for looking thoroughly despite there being so many files. Since CartesianTwist and Wrench are unique messages, they also have unique fields and so should maintain the standard numbering. This is a copying error from CartesianState

domire8 commented on protocol/protobuf/proto/state_representation/space/joint/joint_state.proto:
```diff
@@ -0,0 +1,32 @@
+
+message JointState {
+  State state = 1;
+  repeated string joint_names = 2;
```
I can see why the fields here are of type `repeated`, but my question is if you repeat one field, are the other fields repeated the same number of times? In other words, could you **theoretically** create a message JointState where all those attributes have different length?

eeberhard replied:
Yes, there is no validation between repeated fields, and that's not something the protocol easily supports. However, the JointState constructor and equivalents will complain if the sizes don't match - so a malformed message won't be decoded correctly.

That reminds me that I intend to add another layer of exception handling, so that the `bool decode(const std::string&, T&)` is exception safe even for malformed messages.

domire8 commented on protocol/protobuf/proto/state_representation/space/cartesian/cartesian_state.proto:
```diff
@@ -0,0 +1,46 @@
+import "state_representation/space/spatial_state.proto";
+
+message Vector3d {
+  double x = 1;
```
You don't have to say `required` here?

eeberhard replied:
required and optional are not part of the `proto3` syntax, but nice looking out

domire8 commented on protocol/clproto_cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+if (BUILD_TESTING)
+    file(GLOB_RECURSE TEST_SOURCES test/ test_*.cpp)
+    add_executable(test_${PROJECT_NAME} ${TEST_SOURCES})
+#    target_include_directories(test_${PROJECT_NAME} PUBLIC ${PROJECT_SOURCE_DIR}/include)
```
Can this be removed?

domire8 reviewed (approved):
Well what can I say...Set aside that this is indeed an unfair move in our race and that I'd prefer if you don't exceed the limit of 250 new lines per PR, I think this is really awesome work! We've been talking about this for a long time and now we finally have it and we can start using it cpp / python. I can't wait it to test it myself :smile: Thanks a lot!

I couldn't find any mistakes right now and just for the fact that this PR was created at 4am, you deserve my approval. :rocket:

buschbapti commented on protocol/clproto_cpp/README.md:
```diff
@@ -0,0 +1,47 @@
+
+// The type is automatically derived from the object,
+// but can also be provided explicitly
+clproto::encode<state_representation::CartesianState>(state);
```
does this mean you could encore a `CartesianPose` as a `CartesianState` encoded string? like clproto::encode<CartesianState>(CartesianPose::Random("test"))`? I guess the opposite will throw an encoding error because of some of the extra fields right?

eeberhard replied:
Yes to the first question and no the second. The opposite case is probably more useful:
`clproto::encode<CartesianPose>(CartesianState::Random("test"))`

This is to allow some polymorphism during encoding. If you are upcasting from Pose to State, then message will be sent as a State but the velocity and twist members will be default initialised, which maybe isn't so useful for the recipient. But by "downcast" using the Pose constructor from State, you can choose to send only the Pose data, and the message will explicitly be a Pose message (not decodable as a State). On the receiving end you can always upcast the decoded Pose back to a State object if you want.

buschbapti replied:
Ok this is interesting then. Perfect

buschbapti commented on protocol/clproto_cpp/README.md:
```diff
@@ -0,0 +1,47 @@
+// but can also be provided explicitly
+clproto::encode<state_representation::CartesianState>(state);
+
+
```
uneeded extra line

buschbapti commented on protocol/clproto_cpp/README.md:
```diff
@@ -0,0 +1,47 @@
+
+// The previous method will throw an exception if the
+// message cannot be parsed into that object type.
+// To instead use a 
```
something missing in that comment

buschbapti commented on protocol/clproto_cpp/include/clproto.h:
```diff
@@ -0,0 +1,99 @@
+  JOINT_TORQUES_MESSAGE = 11,
+  SHAPE_MESSAGE = 12,
+  ELLIPSOID_MESSAGE = 13,
+};
```
I see you didn't put `Parameters` yet. Is it harder because those are templates?

eeberhard replied:
Well, despite the large PR, I didn't want to do too much at once. I am not sure yet if encoding parameter messages makes sense, or if indeed this protocol takes care of parameters for us (using string encoded messages as parameters)

buschbapti replied:
It will depend on the API we want to provide the user at the end. If we remove `Parameters` then it means the end-user has to encode decode states from string and handle the `ros2` params calls. The other way is that we keep `Parameters` but I don't think it is actually needed to encode them, just encoding their content should suffice. In `modulo` we can simply encode / decode the contents on `get` and `set` calls.

buschbapti commented on protocol/clproto_cpp/src/clproto.cpp:
```diff
@@ -0,0 +1,732 @@
+ * @param type The state representation StateType
+ * @return The equivalent proto StateType
+ */
+static proto::StateType encoder(const StateType& type) {
```
If I am not mistaken, this `StateType` is a `class enum`. Couldn't we simply encode it as the enum `int` value of each field?

eeberhard replied:
Yes, to be honest I'm not that happy with the StateType enum to begin with. There are two points:

- Whey encode the StateType as an enum message and not an int field:
My intention was that the message is descriptive enough to use without installing control libraries (for later front/backend use). For this, a StateType that is just a number is not helpful, since you don't know what each number means - a named enumeration will provide the type names for the user.

- Why manually assign cases in a switch statement and not just cast from one enum to the other
I could also do a static_cast between enums here, though this requires that the underlying enum index values are consistent between the proto and the control library. It's a little less brittle to assign them explicitly like this.

But, on a larger point, the StateType is a bit messy. It is not granular in the case of frequently used messages (e.g. CARTESIANSTATE for all variants), despite being granular in other cases (PARAMETERS), and contains namespace mismatches (JACOBIANMATRIX for Jacobian, and inconsistent underscores between types).

If clproto constitutes a large enough feature to make a 4.0 release, and would already prompt a refactor of message handling in modulo / downstream users, I would like to refactor this enum in state_representation to be basically equivalent to the StateMessage enum in level of description.

domire8 replied:
4.0 release is gonna be biiiig :rocket:

buschbapti replied:
yes that makes sense

buschbapti commented on protocol/protobuf/proto/state_representation/space/joint/joint_state.proto:
```diff
@@ -0,0 +1,32 @@
+message JointVelocities {
+  State state = 1;
+  repeated string joint_names = 2;
+  repeated double velocities = 4;
```
same comment as @domire8 I guess this should be `3` here and torques in `JointTorques` as well

buschbapti reviewed (commented):
This is super good. Minor comments on the proto messages. The rest is just questions I have. Thanks for this awesome job.

eeberhard commented:
I realised I made some minor changes based on the comments that I never pushed. Please let me know what other pending changes are blocking this PR. There are a couple of smaller PRs I can think of that will follow this one.
- add the clproto_python bindings (which should actually be quite easy because of the way clproto_cpp is encapsulated)
- add CI instruction to run these proto tests
- maybe refactor the StateType
- improve the README 

I also think because of the large size of 4.0 release, and all the holidays at the moment, we should skip a month and postpone the release schedule to end of August.

domire8 commented:
I don't think anything is blocking this PR. Except maybe what you mentioned in the second comment (at least it appears on my github as second one).

> That reminds me that I intend to add another layer of exception handling, so that the bool decode(const std::string&, T&) is exception safe even for malformed messages.

Or do you intend to add that later?

eeberhard commented:
End of week update to this big old PR 

- because I had to rebase against the Dockerfile.base changes, I decided in the end not to have another Dockerfile for the protocol directory. Instead, I added protobuf dependencies to the base development dependencies, which I think makes valid sense. The build time is quite long, but the image is only built on push to main. Conversely, the generated artefacts (binary protoc, and libproto*.a) are not thaaat big, so I think it's ok to keep them around.
- If protobuf libraries are already in the base image, it's easy to add another CI test layer that builds and tests clproto

And, I added that extra exception safety layer to the bool decode() function, so that the documentation is true.

buschbapti commented:
This seems all good. There is just one test failing. Those are new tests I did not see yet.

domire8 commented:
> This seems all good. There is just one test failing. Those are new tests I did not see yet.

This is due to the changes in the Dockerfile.base. Enrico chose to install the protobuf dependencies in the base image but this image is not built and pushed yet (and a workflow dispatch option won't be available until 4.0).
